runtime/compiler/control/CompilationThread.cpp

```diff
@@ -6953,6 +6955,12 @@ TR::CompilationInfoPerThreadBase::preCompilationTasks(J9VMThread * vmThread,
          }
       }
 
+   if (_compInfo.getPersistentInfo()->getJITaaSMode() == CLIENT_MODE && !JITaaS::J9ClientStream::isServerAvailable())
+      {
+      TR_ASSERT(getClientStream() == NULL, "We should not have an open connection when server is unavailable.");
+      setClientStream(JITaaS::J9ClientStream::retryConnectionToServer(OMRPORT_FROM_J9PORT(_jitConfig->javaVM->portLibrary), _compInfo.getPersistentInfo()));
```

Comment:
One disadvantage of this approach is that the client will connect right here even though it may not intend to send a compilation request to the server (maybe because the compilation is cheap and heuristics decide to do it locally).
Also, I would place the check about `JITaaS::J9ClientStream::isServerAvailable()` inside `retryConnectionToServer` (including the assert).

Reply:
It occurred to me that when trying to connect at this point the compilation thread is holding VMaccess. This means that GC cannot happen which in turn mean that the application is going to be blocked.